### PR TITLE
Boundary condition id as list

### PR DESCRIPTION
**Pull request by blaisb**
### Description

One major request of users in the last few years have been to be able to give multiple boundary id when specifying a boundary condition. ( see for example #1331 ). 
 This PR achieves this by parsing a list for the boundary condition and duplicating the boundary condition using all of the ids in the list. This is a simple yet efficient mechanism to enable list for boundary conditions.

I have taken the opportunity to add additional defensive assertions to ensure that the boundary ids are not negative and so on and so forth. This feature is now supported by ALL physics.

### Testing

I have modified an existing test (the cavity one) to use the list specification instead of duplicating the boundary condition subsection. I have also modified the cavity example to use this specification also. I have taken the opportunity to modify one of the tolerances of the cavity example to ensure that the results are sufficiently converged.

### Documentation

I have updated one of the new example to use the new feature, but I don't think this requires specific documentation?

### Checklist (will be removed when merged)
See [this page](https://chaos-polymtl.github.io/lethe/documentation/contributing.html#pull-requests) for more information about the pull request process.

Code related list:
- [x] All in-code documentation related to this PR is up to date (Doxygen format)
- [x] Copyright headers are present and up to date
- [x] Lethe documentation is up to date
- [x] New feature has unit test(s) (preferred) or application test(s), and restart files are in the generator folder
- [x] The branch is rebased onto master
- [x] Changelog (CHANGELOG.md) is up to date
- [x] Code is indented with indent-all and .prm files (examples and tests) with prm-indent

Pull request related list:
- [x] Labels are applied
- [x] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [x] If this PR closes an issue or is related to a project, it is linked in the "Projects" or "Development" section
- [x] If the fix is temporary, an issue is opened
- [x] The PR description is cleaned and ready for merge

### CODE REVIEW

**Comment by blaisb** [comment]
This PR would close #1331

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,12 @@
 All notable changes to the Lethe project will be documented in this file.
 The format is based on [Keep a Changelog](http://keepachangelog.com/).
 
+### [Master] - 2025-09-29
+
+### Added
+
+- MAJOR The id for the boundary condition used to support a single value unsigned integer value, which meant that applying the same boundary condition to multiple ids would require setting up a large amount of boundary condition sections. This PR adds the support to specify a list of boundary id on which a boundary condition is applied. This list of boundary is only used at the level of the parameter file and each id is still considered seperately in the simulation itself. [#1697](https://github.com/chaos-polymtl/lethe/pull/1697)
```
Comment (antonellasuccar) [inline CHANGELOG.md]:
```suggestion
- MAJOR The id for the boundary condition used to support a single unsigned integer value, which meant that applying the same boundary condition to multiple ids would require setting up a large amount of boundary condition sections. This PR adds the support to specify a list of boundary id on which a boundary condition is applied. This list of boundary id is only used at the level of the parameter file and each id is still considered seperately in the simulation itself. [#1697](https://github.com/chaos-polymtl/lethe/pull/1697)
```

File: include/core/boundary_conditions.h [hunk 2]
```diff
@@ -302,127 +305,140 @@ namespace BoundaryConditions
...
+    // Parse the list of boundary ids
+    std::vector<types::boundary_id> boundary_ids =
+      convert_string_to_vector<types::boundary_id>(prm, "id");
+
+    // Check if the list contains at least one boundary id
     AssertThrow(
-      prm.get_integer("id") >= 0,
+      boundary_ids.size() > 0,
       ExcMessage(
         "A boundary id has not been set for one of the fluid dynamics boundary condition. Please ensure that the id is set for every boundary condition."));
```
Comment (antonellasuccar) [inline include/core/boundary_conditions.h]:
```suggestion
        "A boundary id has not been set for one of the fluid dynamics boundary conditions. Please ensure that the id is set for every boundary condition."));
```

File: include/core/boundary_conditions.h [hunk 3]
```diff
@@ -302,127 +305,140 @@ namespace BoundaryConditions
...
+          this->type[boundary_id] = BoundaryType::slip;
+        if (op == "function" || op == "function weak" || op == "partial slip")
+          {
+            if (op == "function")
+              this->type[boundary_id] = BoundaryType::function;
+            else if (op == "partial slip")
+              this->type[boundary_id] = BoundaryType::partial_slip;
+            else
+              this->type[boundary_id] = BoundaryType::function_weak;
+          }
```
Comment (antonellasuccar) [inline include/core/boundary_conditions.h:375]: Are we using this if structure because there are a lot of cases where op=="function" is true?

File: include/core/boundary_conditions.h [hunk 4]
```diff
@@ -666,81 +682,92 @@ namespace BoundaryConditions
...
+    std::vector<types::boundary_id> boundary_ids =
+      convert_string_to_vector<types::boundary_id>(prm, "id");
+
+    // Check if the list contains at least one boundary id
     AssertThrow(
-      prm.get_integer("id") >= 0,
+      boundary_ids.size() > 0,
       ExcMessage(
-        "An invalid boundary id has been given for a heat transfer boundary condition."));
+        "A boundary id has not been set for one of the fluid dynamics boundary condition. Please ensure that the id is set for every boundary condition."));
```
Comment (antonellasuccar) [inline include/core/boundary_conditions.h]:
```suggestion
        "A boundary id has not been set for one of the heat transfer boundary conditions. Please ensure that the id is set for every boundary condition."));
```

File: include/core/boundary_conditions.h [hunk 5]
```diff
@@ -894,51 +921,62 @@ namespace BoundaryConditions
...
+
+    // Check if the list contains at least one boundary id
     AssertThrow(
-      prm.get_integer("id") >= 0,
+      boundary_ids.size() > 0,
       ExcMessage(
-        "An invalid boundary id has been given for a tracer boundary condition."));
-
-    types::boundary_id boundary_id = prm.get_integer("id");
+        "A boundary id has not been set for one of the fluid dynamics boundary condition. Please ensure that the id is set for every boundary condition."));
```
Comment (antonellasuccar) [inline include/core/boundary_conditions.h]:
```suggestion
        "A boundary id has not been set for one of the tracer boundary conditions. Please ensure that the id is set for every boundary condition."));
```

File: include/core/boundary_conditions.h [hunk 6]
```diff
@@ -1097,69 +1135,77 @@ namespace BoundaryConditions
...
-    AssertThrow(this->type.find(boundary_id) == this->type.end(),
-                CahnHilliardBoundaryDuplicated(boundary_id));
-
-    // Create and parse the phase order boundary condition for all cases.
-    prm.enter_subsection("phi");
-    this->bcFunctions[boundary_id] =
-      std::make_shared<CahnHilliardBoundaryFunctions<dim>>();
-    bcFunctions[boundary_id]->phi.parse_parameters(prm);
-    prm.leave_subsection();
+        "A boundary id has not been set for one of the fluid dynamics boundary condition. Please ensure that the id is set for every boundary condition."));
```
Comment (antonellasuccar) [inline include/core/boundary_conditions.h]:
```suggestion
        "A boundary id has not been set for one of the Cahn Hilliard boundary conditions. Please ensure that the id is set for every boundary condition."));
```

File: include/core/boundary_conditions.h [hunk 7]
```diff
@@ -1307,45 +1353,53 @@ namespace BoundaryConditions
...
-      prm.get_integer("id") >= 0,
+      boundary_ids.size() > 0,
       ExcMessage(
-        "An invalid boundary id has been given for a VOF boundary condition."));
-
-    types::boundary_id boundary_id = prm.get_integer("id");
-
-    AssertThrow(this->type.find(boundary_id) == this->type.end(),
-                VOFBoundaryDuplicated(boundary_id));
+        "A boundary id has not been set for one of the fluid dynamics boundary condition. Please ensure that the id is set for every boundary condition."));
```
Comment (antonellasuccar) [inline include/core/boundary_conditions.h]:
```suggestion
        "A boundary id has not been set for one of the VOF boundary conditions. Please ensure that the id is set for every boundary condition."));
```

File: include/core/boundary_conditions.h [hunk 8]
```diff
@@ -302,127 +305,140 @@ namespace BoundaryConditions
...
-
-        // We attribute a periodic neighbor boundary type to the neighbor
-        // boundary to ensure that all boundaries have a defined type
-        this->type[periodic_boundary_id] = BoundaryType::periodic_neighbor;
-
-        // We store the periodic id and direction
-        this->periodic_neighbor_id[boundary_id] = periodic_boundary_id;
-        this->periodic_direction[boundary_id] =
-          prm.get_integer("periodic_direction");
+        AssertThrow(this->type.find(boundary_id) == this->type.end(),
```
Comment (antonellasuccar) [inline include/core/boundary_conditions.h:323]: This tests for duplicate boundary ids. Now sure I see where we are testing if the boundary ids are positive.

File: include/core/boundary_conditions.h (hunk 3 above)
Comment (blaisb) [reply include/core/boundary_conditions.h:375]: No. There's no real reason why we are using this if structure here. This part of the code only runs once at the beginning so there's no performance in mind here. Mostly it's like that for history reason I presume

File: include/core/boundary_conditions.h (hunk 8 above)
Comment (blaisb) [reply include/core/boundary_conditions.h:323]: Yeah that's a bad comment. We test for the positivity when we parse the id themselves now.

File: include/core/boundary_conditions.h [hunk 9]
```diff
@@ -302,127 +305,140 @@ namespace BoundaryConditions
...
-    // Establish the type of boundary condition
-    const std::string op = prm.get("type");
-    if (op == "none")
-      this->type[boundary_id] = BoundaryType::none;
-    if (op == "noslip")
-      this->type[boundary_id] = BoundaryType::noslip;
-    if (op == "slip")
-      this->type[boundary_id] = BoundaryType::slip;
-    if (op == "function" || op == "function weak" || op == "partial slip")
+    // Loop through all boundary ids to ensure that they are all non-negative
```
Comment (blaisb) [inline include/core/boundary_conditions.h]:
```suggestion
```

File: include/core/boundary_conditions.h [hunk 10]
```diff
@@ -4,6 +4,10 @@
 #ifndef lethe_boundary_conditions_h
 #define lethe_boundary_conditions_h
 
+#include <core/utilities.h>
+
+#include "dem/find_boundary_cells_information.h"
```
Comment (voferreira) [inline include/core/boundary_conditions.h]: Should this be added?

File: include/core/boundary_conditions.h [hunk 11]
```diff
@@ -4,6 +4,10 @@
 #ifndef lethe_boundary_conditions_h
```
Comment (voferreira) [inline include/core/boundary_conditions.h:4]: Should the copyright be updated?

File: include/core/boundary_conditions.h (hunk 10 above)
Comment (blaisb) [reply include/core/boundary_conditions.h]: Not the DEM that's for sure that's a mistake on my part. Good catch

File: include/core/boundary_conditions.h (hunk 11 above)
Comment (blaisb) [reply include/core/boundary_conditions.h:4]: Yup I will do so

**Comment by blaisb** [comment]
@antonellasuccar I'll wait for your final OK before merging (always put a green checkmark if you don't want to "re-review" the PR please :) )

File: include/core/boundary_conditions.h [hunk 12]
```diff
@@ -1,13 +1,11 @@
-// SPDX-FileCopyrightText: Copyright (c) 2019-2024 The Lethe Authors
+// SPDX-FileCopyrightText: Copyright (c) 2019-2054 The Lethe Authors
```
Comment (voferreira) [inline include/core/boundary_conditions.h]: Small typo here :)

Comment (voferreira) [inline include/core/boundary_conditions.h]:
```suggestion
// SPDX-FileCopyrightText: Copyright (c) 2019-2025 The Lethe Authors
```

File: applications_tests/lethe-fluid-particles/generators/gas_solid_fluidized_bed_generator_step_1_dem.prm
Comment (AmishgaAlphonius) [inline applications_tests/lethe-fluid-particles/generators/gas_solid_fluidized_bed_generator_step_1_dem.prm:1]: copyright update needs to be updated here

File: applications_tests/lethe-fluid/cavity_adjoint_gls.prm
Comment (AmishgaAlphonius) [inline applications_tests/lethe-fluid/cavity_adjoint_gls.prm:1]: copyright here also

File: examples/incompressible-flow/2d-lid-driven-cavity/Reynolds_7500/cavity.prm
Comment (AmishgaAlphonius) [inline examples/incompressible-flow/2d-lid-driven-cavity/Reynolds_7500/cavity.prm:1]: copyright here also

File: examples/incompressible-flow/2d-lid-driven-cavity/cavity.prm
Comment (AmishgaAlphonius) [inline examples/incompressible-flow/2d-lid-driven-cavity/cavity.prm:1]: here also

File: include/core/utilities.h [hunk 13]
```diff
@@ -946,6 +946,62 @@ create_random_number_container(std::vector<double> &random_container,
...
+/**
+ * @brief Return the vector of size N of entry @entry_string. If the entry is
+ * specified in the .prm file, then the changed value is returned, otherwise
+ * the default value is returned. If the entry is not equivalent to a vector,
+ * an error will be thrown.
+ *
+ * @tparam T The type of vector (int, unsigned int or double)
+ * @param prm A parameter handler which is currently used to parse the simulation
+ * information.
+ * @param entry_string A declare string in the parameter file.
```
Comment (AmishgaAlphonius) [inline include/core/utilities.h]:
```suggestion
 * @param[in] prm A parameter handler which is currently used to parse the simulation
 * information.
 * @param[in] entry_string A declare string in the parameter file.
```

File: include/core/utilities.h [hunk 14]
```diff
@@ -946,6 +946,62 @@ create_random_number_container(std::vector<double> &random_container,
...
+ *
+ * @return A std::vector<double> corresponding to the entry_string in the prm file.
+ */
+template <typename T>
+std::vector<T>
+convert_string_to_vector(ParameterHandler &prm, const std::string &entry_string)
```
Comment (AmishgaAlphonius) [inline include/core/utilities.h]: Here, shouldn't `prm` be passed as `const`?
